libsplinter/src/service/scabbard/rest_api.rs

```diff
@@ -227,6 +227,102 @@ pub fn make_get_batch_status_endpoint() -> ServiceEndpoint {
+    ServiceEndpoint {
+        service_type: SERVICE_TYPE.into(),
+        route: "/state/{address}".into(),
+        method: Method::Get,
+        handler: Arc::new(move |request, _, service| {
+            let scabbard = match service.as_any().downcast_ref::<Scabbard>() {
+                Some(s) => s,
+                None => {
+                    return Box::new(HttpResponse::InternalServerError().json(json!({
+                        "message": "service is not Scabbard"
+                    })).into_future())
+                }
+            };
+
+            let address = request
+                .match_info()
+                .get("address")
+                .expect("address should not be none");
+
+            Box::new(match scabbard.get_state_at_address(address) {
+                Ok(Some(value)) => HttpResponse::Ok().json(value).into_future(),
+                Ok(None) => HttpResponse::NotFound().finish().into_future(),
+                Err(err) => HttpResponse::InternalServerError()
+                    .json(json!({
+                        "message": err.to_string()
+                    }))
+                    .into_future(),
+            })
+        }),
+    }
+}
+
+#[cfg(feature = "scabbard-get-state")]
+pub fn make_get_state_with_prefix_endpoint() -> ServiceEndpoint {
+    ServiceEndpoint {
+        service_type: SERVICE_TYPE.into(),
+        route: "/state".into(),
+        method: Method::Get,
+        handler: Arc::new(move |request, _, service| {
+            let scabbard = match service.as_any().downcast_ref::<Scabbard>() {
+                Some(s) => s,
+                None => {
+                    return Box::new(HttpResponse::InternalServerError().json(json!({
+                        "message": "service is not Scabbard"
+                    })).into_future())
+                }
+            };
+
+            let query: web::Query<HashMap<String, String>> =
+                if let Ok(q) = web::Query::from_query(request.query_string()) {
+                    q
+                } else {
+                    return Box::new(
+                        HttpResponse::BadRequest()
+                            .json(json!({
+                                "message": "Invalid query"
+                            }))
+                            .into_future(),
+                    );
+                };
+
+            let prefix = query.get("prefix").map(String::as_str);
+
+            Box::new(match scabbard.get_state_with_prefix(prefix) {
+                Ok(state_iter) => {
+                    let res = state_iter
+                        .map(|res| {
+                            res.map(|(address, value)| {
+                                json!({
+                                    "address": address,
+                                    "value": value,
+                                })
+                            })
+                        })
+                        .collect::<Result<Vec<_>, _>>();
```

Comment:
This doesn't currently support paging through the state entries because that isn't supported by transact.

Reply:
Mainly, this is due to the fact that it is currently difficult to find a starting point for a leaf, but I have some ideas for that in transact.

Reply:
By adding more meta information to the tree nodes?

Reply:
I don't think that's needed, but I think some simple calculations can be made based on an address to find where to start in the tree.  Most of the paging in sawtooth initially relied on a  position in the result set.  This was simple, and allowed users to create the page links without any knowledge of the underlying system.  In order to actually optimize these queries, the "next" item needs to really be specific to the store being iterated over.  A good example of this was the change in block iteration, where it uses a representation of block height for iteratation.  

For state, we could use the combination of `(state root, Option<Subtree>, address)` as the location in the iterator.  This can get us a starting point in the tree, and iterate over the leaves from there.

Reply:
In order to implement this, we need to implement some changes in [hyperledger/transact](https://github.com/hyperledger/transact) first.